### PR TITLE
Third set of Hackerrank excercises and attendance - Brandon Beltran

### PR DESCRIPTION
**Pull request by Armastardo**
My attendance for August 27th was lost along the way, some git fun!

### CODE REVIEW

File: BrandonBeltran/hackerrank/03/JavaHashset.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/* This solution is the one I came on my own without Hackerrank's code.
+*  To be honest, I like it better, I include it for the sake of completion*/
+
+public class JavaHashset {
+    public static void main(String[] args) {
+
+        Scanner scanner = new Scanner(System.in);
+        int sets = Integer.parseInt(scanner.nextLine());
+
+        HashSet<String> names = new HashSet();
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaHashset.java:15]: Code for Interface, not for Implementation
https://www.javaguides.net/2018/06/java-collection-framework-best-practices.html

File: BrandonBeltran/hackerrank/03/JavaHashsetHackerrank.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        int queries = scanner.nextInt();
+        String [] pairLeft = new String[queries];
+        String [] pairRight = new String[queries];
+
+        for (int index = 0; index < queries; index++) {
+            pairLeft[index] = scanner.next();
+            pairRight[index] = scanner.next();
+        }
+
+        HashSet<String> pairs = new HashSet<>();
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaHashsetHackerrank.java:22]: Code for Interface, not for Implementation
https://www.javaguides.net/2018/06/java-collection-framework-best-practices.html

File: BrandonBeltran/hackerrank/03/JavaList.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        int queries = scanner.nextInt();
+
+        for(int iterator = 0; iterator < queries; iterator++){
+            String operation = scanner.next();
+            int position = scanner.nextInt();
+
+            if(operation.equals("Insert")){
+                int addition = scanner.nextInt();
+                numbers.add(position, addition);
+            }else if (operation.equals("Delete")){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaList.java:27]: What if position > number.size() when trying numbers.remove() ???

File: BrandonBeltran/hackerrank/03/JavaList.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+                numbers.remove(position);
+            }
+        }
+
+        for(int index = 0; index < numbers.size(); index++){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaList.java:32]: for... if... if... for..... what about something like:

while (iterator.hasNext()) {
}

File: BrandonBeltran/hackerrank/03/JavaMap.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+
+public class JavaMap {
+
+    public static void main(String[] args) {
+
+        Map<String, Integer> phoneBook = new HashMap<>();
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaMap.java:11]: Generics!! Great!!

File: BrandonBeltran/hackerrank/03/JavaMap.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+            String name = scanner.nextLine();
+            int number = Integer.parseInt(scanner.nextLine());
+            phoneBook.put(name, number);
+
+        }
+
+        while(scanner.hasNext()){
+            String query = scanner.nextLine();
+            if(phoneBook.get(query) != null){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaMap.java:25]: what about map.contains() instead if(phoneBook.get() != null) ??

File: BrandonBeltran/hackerrank/03/JavaSort.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
+package bootcamp;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Scanner;
+
+class Student implements Comparable<Student>{
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaSort.java:8]: Comparable!! [clapping]
But implements compareTo just for one double field ?? Just let's keep that in mind for future coding

File: BrandonBeltran/hackerrank/03/JavaSort.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            String fname = scanner.next();
+            double cgpa = scanner.nextDouble();
+
+            Student st = new Student(id, fname, cgpa);
+            studentList.add(st);
+
+            testCases--;
+        }
+
+        Collections.sort(studentList);
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaSort.java:60]: Collections.sort() Awesome!!

File: BrandonBeltran/hackerrank/03/JavaSort.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+        if(comparison == 0){
+            comparison = this.fname.compareTo(versus.getFname());
+        }
+
+        return comparison;
+    }
+}
+
+public class JavaSort {
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaSort.java:41]: Each class must be in one file please

File: BrandonBeltran/hackerrank/03/RansomNote.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Stream;
+import static java.util.stream.Collectors.toList;
+
+class Result {
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/RansomNote.java:11]: Each class in one file please

File: BrandonBeltran/hackerrank/03/RansomNote.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    public static void checkMagazine(List<String> magazine, List<String> note) {
+        HashMap<String, Integer> magazineWords = new HashMap<>();
+
+        for(String word: magazine){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/RansomNote.java:16]: Two foor loop almost identical... hmmmmmmmmm
Maybe another method that receives a list and an operation ? IDK
If you have another option, please let me know

File: BrandonBeltran/hackerrank/03/RansomNote.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+}
+
+public class RansomNote {
+    public static void main(String[] args) throws IOException {
+
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
+
+        String[] totalWords = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/RansomNote.java:44]: What is "\\s+$" ??? Is used 3 times in the code... Maybe if you define a constant with a meaningful name we all can get it right?

**Review by luisod78 — CHANGES_REQUESTED** [review]

File: BrandonBeltran/hackerrank/03/JavaList.java (hunk 3 above)
Comment (Armastardo) [reply BrandonBeltran/hackerrank/03/JavaList.java:27]: I'd throw an IndexOutOfBoundsException!! I'll fix that, thank you :)

File: BrandonBeltran/hackerrank/03/RansomNote.java (hunk 11 above)
Comment (Armastardo) [reply BrandonBeltran/hackerrank/03/RansomNote.java]: I moved things a little bit, please let me know how this solutions looks!

File: BrandonBeltran/hackerrank/03/RansomNote.java (hunk 12 above)
Comment (Armastardo) [reply BrandonBeltran/hackerrank/03/RansomNote.java]: Made a final string called WHITESPACE_AND_EOL, so people know what that small regex means. :)

File: BrandonBeltran/hackerrank/03/JavaArrayList.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
+        int collections = scanner.nextInt();
+        ArrayList <List> numbers = new ArrayList<>();
+
+        for(int collectionIterator = 0; collectionIterator < collections; collectionIterator++){
+            int size = scanner.nextInt();
+
+            ArrayList <Integer> numberSubList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (luisod78) [reply BrandonBeltran/hackerrank/03/JavaArrayList.java:17]: Resolved??? numberSubList still  ArrayList

File: BrandonBeltran/hackerrank/03/JavaList.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            numbers.add(scanner.nextInt());
+        }
+
+        int queries = scanner.nextInt();
+
+        for(int iterator = 0; iterator < queries; iterator++){
+            String operation = scanner.next();
+            int position = scanner.nextInt();
+
+            if(position <= numbers.size() || position < 0){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaList.java:25]: Great job!

File: BrandonBeltran/hackerrank/03/JavaList.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+                    default:
+                        break;
+                }
+            }else{
+                System.out.println("Error! Invalid position.");
+            }
+        }
+
+        Iterator numbersIterator = numbers.iterator();
+        while(numbersIterator.hasNext()){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaList.java:43]: cool

File: BrandonBeltran/hackerrank/03/JavaMap.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+            String name = scanner.nextLine();
+            int number = Integer.parseInt(scanner.nextLine());
+            phoneBook.put(name, number);
+
+        }
+
+        while(scanner.hasNext()){
+            String query = scanner.nextLine();
+            if(phoneBook.containsKey(query)){
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/JavaMap.java:25]: perfect!! don't forget Collections has done -almost- all work for us, sometimes you only need to find it in API doc

File: BrandonBeltran/hackerrank/03/RansomNote.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.List;
+import java.util.stream.Stream;
+import static java.util.stream.Collectors.toList;
+
+public class RansomNote {
+
+    public static final String WHITESPACE_AND_EOL = "\\s+$";
```
Comment (luisod78) [inline BrandonBeltran/hackerrank/03/RansomNote.java:12]: now we are talkin'
code more readable
[clapping]

**Review by luisod78 — APPROVED** [review]
Great job refactoring 
Maybe you'd like to google: How to convert List to Map?
just for fun

**Review by luisalanmz — APPROVED** [review]
Ok

**Pull request merged**